apidef/oas/middleware.go

```diff
@@ -802,3 +819,64 @@ func (ap *AuthenticationPlugin) ExtractTo(api *apidef.APIDefinition) {
 	api.CustomMiddleware.AuthCheck.Path = ap.Path
 	api.CustomMiddleware.AuthCheck.RawBodyOnly = ap.RawBodyOnly
 }
+
+// CustomPlugin configures custom plugin.
+type CustomPlugin struct {
+	// Enabled enables the custom pre plugin.
+	Enabled bool `bson:"enabled" json:"enabled"` // required.
+	// FunctionName is the name of authentication method.
+	FunctionName string `bson:"functionName" json:"functionName"` // required.
+	// Path is the path to shared object file in case of gopluign mode or path to js code in case of otto auth plugin.
+	Path string `bson:"path" json:"path"` // required.
+	// RawBodyOnly if set to true, do not fill body in request or response object.
+	RawBodyOnly bool `bson:"rawBodyOnly,omitempty" json:"rawBodyOnly,omitempty"`
+}
+
+// CustomPlugins is a list of CustomPlugin.
+type CustomPlugins []CustomPlugin
+
+// Fill fills CustomPlugins from supplied Middleware definitions.
+func (c CustomPlugins) Fill(mwDefs []apidef.MiddlewareDefinition) {
+	for i, mwDef := range mwDefs {
+		c[i] = CustomPlugin{
+			Enabled:      !mwDef.Disabled,
+			Path:         mwDef.Path,
+			FunctionName: mwDef.Name,
+			RawBodyOnly:  mwDef.RawBodyOnly,
+		}
+	}
+}
+
+// ExtractTo extracts CustomPlugins into supplied Middleware definitions.
+func (c CustomPlugins) ExtractTo(mwDefs []apidef.MiddlewareDefinition) {
+	for i, plugin := range c {
+		mwDefs[i] = apidef.MiddlewareDefinition{
+			Disabled:    !plugin.Enabled,
+			Name:        plugin.FunctionName,
+			Path:        plugin.Path,
+			RawBodyOnly: plugin.RawBodyOnly,
+		}
+	}
+}
+
+// PrePlugin configures pre stage plugins.
+type PrePlugin struct {
+	// Plugins configures custom plugins to be run on pre authentication stage.
+	// The plugins would be executed in the order of configuration in the list.
+	Plugins CustomPlugins `bson:"plugins" json:"plugins"`
+}
+
+// Fill fills PrePlugin from supplied Tyk classic api definition.
+func (p *PrePlugin) Fill(api apidef.APIDefinition) {
+	p.Plugins = make(CustomPlugins, len(api.CustomMiddleware.Pre))
```

Comment:
we can avoid shouldomit here with a simple length check and simplify the code for this function; if the length is zero, p.Plugins = nil and return, otherwise .Fill

Reply:
I had thought about this, added should omit to be consistent with other functions. len check is much cleaner and doesn't allocate if not required.